executor/foreign_key.go

```diff
@@ -727,120 +718,100 @@ func (fkc *FKCascadeExec) fetchUpdatedValuesCouple() *UpdatedValuesCouple {
-func GenCascadeUpdateSQL(schema, table, idx model.CIStr, fk *model.FKInfo, couple *UpdatedValuesCouple) (string, error) {
-	buf := bytes.NewBuffer(nil)
-	buf.WriteString("UPDATE `")
-	buf.WriteString(schema.L)
-	buf.WriteString("`.`")
-	buf.WriteString(table.L)
-	buf.WriteString("`")
-	if idx.L != "" {
-		// Add use index to make sure the optimizer will use index instead of full table scan.
-		buf.WriteString(" USE INDEX(`")
-		buf.WriteString(idx.L)
-		buf.WriteString("`)")
-	}
-	buf.WriteString(" SET ")
-	for i, col := range fk.Cols {
-		if i > 0 {
-			buf.WriteString(", ")
-		}
-		buf.WriteString("`" + col.L)
-		buf.WriteString("` = ")
-		val, err := genFKValueString(couple.NewValues[i])
-		if err != nil {
-			return "", err
+// GenCascadeUpdateAST uses to generate cascade update ast, export for test.
+func GenCascadeUpdateAST(schema, table, idx model.CIStr, cols []*model.ColumnInfo, couple *UpdatedValuesCouple) *ast.UpdateStmt {
+	list := make([]*ast.Assignment, 0, len(cols))
+	for i, col := range cols {
+		v := &driver.ValueExpr{Datum: couple.NewValues[i]}
+		v.Type = col.FieldType
+		assignment := &ast.Assignment{
+			Column: &ast.ColumnName{Name: col.Name},
+			Expr:   v,
 		}
-		buf.WriteString(val)
+		list = append(list, assignment)
 	}
-	err := genCascadeSQLWhereCondition(buf, fk, couple.OldValuesList)
-	if err != nil {
-		return "", err
+	updateStmt := &ast.UpdateStmt{
+		TableRefs: genTableRefsAST(schema, table, idx),
+		Where:     genWhereConditionAst(cols, couple.OldValuesList),
+		List:      list,
 	}
-	return buf.String(), nil
+	return updateStmt
 }
 
-func genCascadeSQLWhereCondition(buf *bytes.Buffer, fk *model.FKInfo, fkValues [][]types.Datum) error {
-	buf.WriteString(" WHERE (")
-	for i, col := range fk.Cols {
-		if i > 0 {
-			buf.WriteString(", ")
-		}
-		buf.WriteString("`" + col.L + "`")
-	}
-	buf.WriteString(") IN (")
-	for i, vs := range fkValues {
-		if i > 0 {
-			buf.WriteString(", (")
-		} else {
-			buf.WriteString("(")
-		}
-		for i := range vs {
-			val, err := genFKValueString(vs[i])
-			if err != nil {
-				return err
-			}
-			if i > 0 {
-				buf.WriteString(",")
+func genTableRefsAST(schema, table, idx model.CIStr) *ast.TableRefsClause {
+	tn := &ast.TableName{Schema: schema, Name: table}
+	if idx.L != "" {
+		tn.IndexHints = []*ast.IndexHint{{
+			IndexNames: []model.CIStr{idx},
+			HintType:   ast.HintUse,
+			HintScope:  ast.HintForScan,
+		}}
+	}
+	join := &ast.Join{Left: &ast.TableSource{Source: tn}}
+	return &ast.TableRefsClause{TableRefs: join}
+}
+
+func genWhereConditionAst(cols []*model.ColumnInfo, fkValues [][]types.Datum) ast.ExprNode {
+	if len(cols) > 1 {
+		return genWhereConditionAstForMultiColumn(cols, fkValues)
+	}
+	valueList := make([]ast.ExprNode, 0, len(fkValues))
+	for _, fkVals := range fkValues {
+		v := &driver.ValueExpr{Datum: fkVals[0]}
+		v.Type = cols[0].FieldType
+		row := &ast.ParenthesesExpr{Expr: v}
```

Comment:
can we eliminate the `ParenthesesExpr` wrapper?

Reply:
Nice catch,  done.